### ISSUE TITLE
False positives and negatives when vendor-prefixed values are used in function-linear-gradient-no-nonstandard-direction

### ISSUE THREAD

**Issue by pocke** [stylelint/stylelint#1939]
> Describe the issue. Is it a bug or a feature request (new rule, new option, etc.)?

A bug.

If `linear-gradient` function is with a vendor prefix, `function-linear-gradient-no-nonstandard-direction` rule has false positives and false negatives.

In standard CSS, `linear-gradient` expects `to` prefix for direction. However, with vendor prefix, `to` prefix is not acceptable as direction.

> The -moz- prefixed rules are used for compatibility with older versions (Fx 3.6 to Fx 15). The -webkit-prefix is the only one that still needs to be included for Android 4.3-, iOS up to 6.1, and Safari 6. When using a prefix, do not use 'to'.
>
> https://developer.mozilla.org/en-US/docs/Web/CSS/linear-gradient#Cross-browser_gradients

So, this rule is incorrect when the function is with a vendor prefix.

> Which rule, if any, is this issue related to?

`function-linear-gradient-no-nonstandard-direction`

> What CSS is needed to reproduce this issue?

```css
.foo {
  // for false positives
  background: -webkit-linear-gradient(top, #fff, #000); 
  background:    -moz-linear-gradient(top, #fff, #000); 
  background:      -o-linear-gradient(top, #fff, #000); 

  // for false negatives
  background: -webkit-linear-gradient(to bottom, #fff, #000 );
  background:    -moz-linear-gradient(to bottom, #fff, #000 );
  background:      -o-linear-gradient(to bottom, #fff, #000 );
}
```

> What stylelint configuration is needed to reproduce this issue?

```json
{
  "rules": {
    "function-linear-gradient-no-nonstandard-direction": true
  }
}
```

> Which version of stylelint are you using?

`7.3.1`

> How are you running stylelint: CLI, PostCSS plugin, Node API?

CLI interface.
e.g. `stylelint test.css`

> Does your issue relate to non-standard syntax (e.g. SCSS, nesting, etc.)?

No, I can reproduce the issue with standard syntax.
However, I've found the issue when I use less and autoprefixer.

> What did you expect to happen?

No warnings to be flagged with "for false positives" in the above css.
And I expect warnings with "for false negatives" in the above css.

> What actually happened (e.g. what warnings or errors you are getting)?

In "for false positives", the following warnings were flagged: 

```shell
test.css
 3:39  ✖  Unexpected nonstandard direction   function-linear-gradient-no-nonstandard-direction
 4:39  ✖  Unexpected nonstandard direction   function-linear-gradient-no-nonstandard-direction
 5:39  ✖  Unexpected nonstandard direction   function-linear-gradient-no-nonstandard-direction
```

And I can't get any warnings in "for false negatives" properties.

**Comment by jeddy3** [comment]
@pocke Thank you for an excellent issue write-up and for using the issue template!

> However, I've found the issue when I use less and autoprefixer.

This does appear to be a bug, and one that hasn't been reported before. I think this is because stylelint is usually used to lint source code *before* autoprefixer performs any transformations i.e. the source code usually contains only *non-standard* prefixed things like `-webkit-font-smoothing`, rather than prefixed *standard* things like `linear-gradient`. 

Unless you have a specific reason not to, we recommend linting your source code before transforming it. 

Having said that, we know of *some* teams that do have specific reasons to lint their compiled CSS (which would likely include autoprefixed code).

As such, I think your fix is valid and should be included for this particular *edge-case*. 

So, thanks for that :)
